content/blog/2019-11-13-loop-perforation/index.md

```diff
@@ -0,0 +1,420 @@
+For these reasons, we enter a total accuracy score for each error metric that we collect, as well as for each of many variances.
+We then use all of these metricscass to compute the Pareto frontier.
+
+These error metrics are more finicky than they might appear for real programs.
+Consider a Sobel image filter for edge detection.
+We use an implementation of this filter from the [ACCEPT benchmarks][] for approximate computing.
+
+Running the standard implementation on this image of cute buffalo:
+
+<img src="sobel-input.png" width=500/>
+
+Gives the following correct output:
+
+<img src="sobel-output.png" width=500/>
+
+Running our driver with a default configuration perforates the loops in this application at the following rates:
+
+```
+{
+    \"load_image_data\": {\"%81,%84,%85,%97,%98,%88,%95\": 5, \"%85,%88,%95\": 5},
+    \"save_image_data\": {\"%11,%14,%15,%28,%29,%18,%26\": 1, \"%15,%18,%26\": 1},
+    \"sobel_filtering\": {\"%74,%77,%78,%88,%89,%81,%86\": 5, \"%78,%81,%86\": 5}
+}
+```
+
+The driver chooses this configuration because one of the error metrics, `"l2_100000"`, only has an error of 0.185.
+However the resulting image looks...very bad.
+
+
+<img src="sobel-perforated-default.png" width=500/>
+
+With a little design space exploration, we can try and find other metrics that produce less psychedelic results.
+The following result only considers the `l1_10000` metric and allows an error of 0.6.
+
+<img src="sobel-perforated-l1-10000-0.6.png" width=500/>
+
+This image is _kind of_ on the right track, in that we can clearly see some of the edges in the original image (just shifted and stretched to be almost useless).
+Remarkably, the above image was the closest to correct we could find with any loop perforation at all via a casual design space exploration.
+However, this is clearly **not** an acceptable result for edge detection!
+
+Ultimately, this example illustrates that finding the right error metric might be both difficult and counterintuitive—to get a better image that gained any speedup, we had to select a higher error rate, just for a different metric.
+
+
+[paper]: https://dl.acm.org/citation.cfm?id=2025133
+
+### Tests and Benchmarks
+
+We implemented three small test programs:
+1. `sum-to-n`: Sums all numbers between 1 and n.
+2. `alloc-loop`: A program that performs arithmetic on an array of integer pointers.
+3. `matrix_multiplication`: Matrix multiplication of two random 100-by-100 matrices.
+
+We additionally ran loop perforation on three larger benchmark programs:
+1. `sobel`: A Sobel filter from the [ACCEPT benchmarks][] for approximate computing.
+2. `img-blur`: A Gaussian blur that operates on the same inputs as `sobel`; we implemented this ourselves using Sobel as a model. They share PGM processing code.
+3. `blackscholes`: From the [PARSEC][] benchmark suite. We chose this benchmark for ease and simplicity of compilation, as well as having a straightforward potential for error metrics for the output data.
```

Comment:
If I recall correctly, one of the version of the original loop perforation paper that's out there hilariously perforates the outermost loop in blackscholes—the one that runs the entire identical computation several times! To everyone's shock, you can skip iterations of this loop while losing no accuracy whatsoever.